File: JeongGod/5week/programmers/92341.py
import math


def calc(exceed_time : int) -> int:
    # 초과하지 않았다면
    if exceed_time <= 0:
        return 0
    return math.ceil(exceed_time / i_time) * i_price

def hour_to_minute(time : str):
    hour, minute = map(int, time.split(":"))
    return hour*60 + minute

def solution(fees, records):
    global b_time, b_price, i_time, i_price
    """
    1. 입차된 후에 출차 내역이 없다면 23:59분에 출차한 것으로
    2. 기본 요금, 기본 요금 + 단위 시간(올림) * 단위 요금
    차량 번호가 작은 자동차부터 청구할 주차 요금
    """
    answer = []
    b_time, b_price, i_time, i_price = fees
    cars = dict()
    for time, car, com in map(lambda x:x.split(), records):
        if com == "IN":
            target = cars.get(car)
            # 처음 차량이 들어왔다면
            if target is None:
                cars[car] = [hour_to_minute(time), 0, True]
                continue
            # 다시 들어온다면
            target[0] = hour_to_minute(time)
            target[2] = True
        else:
            target = cars.get(car)
            target[1] += hour_to_minute(time) - target[0]
            target[2] = False
            
    for car in sorted(cars.keys()):
        # 출차하지 않은 경우
        target = cars[car]
        if target[2] == 1:
            target[1] += hour_to_minute("23:59") - target[0]
        # 계산
        answer.append(b_price + calc(target[1] - b_time))
        
    return answer

File: jekyun-park/week5/programmers/92341.swift
//
//  main.swift
//  SwiftAlgorithms
//
//  Created by 박제균 on 2022/01/28.
//

import Foundation

func solution(_ fees: [Int], _ records: [String]) -> [Int] {

    var timeRecord: [String: [[String]]] = [:]
    var feeDictionary: [String:Int] = [:]


    for record in records {
        let log = record.split(separator: " ")
        let time = String(log[0])
        let car = String(log[1])
//        let inOut = log[2]

        if timeRecord[car] == nil {
            timeRecord[car] = []
            timeRecord[car]?.append([time])
        } else {
            if timeRecord[car]![timeRecord[car]!.count - 1].count == 2 {
                timeRecord[car]!.append([time])
            } else {
                timeRecord[car]![timeRecord[car]!.count - 1].append(time)
            }
        }
    }

    for log in timeRecord {
        if log.value.last!.count % 2 != 0 {
            timeRecord[log.key]![timeRecord[log.key]!.count - 1].append("23:59")
        }
    }
    
    
    
    for car in timeRecord {
        var sum = 0
        timeRecord[car.key]!.forEach {
            sum += calculateTime($0[0],$0[1])
        }
        feeDictionary[car.key] = calculateFee(sum, fees)
    }

    var answer:[Int] = []
    feeDictionary.sorted(by: <).forEach { answer.append($0.value) }

    return answer
}




func calculateFee(_ parkTime: Int, _ fees: [Int]) -> Int {
    
    let standardTime = fees[0]
    let standardFee = fees[1]
    let unitTime = fees[2]
    let unitFee = fees[3]
    if parkTime <= standardTime { return standardFee }
    let totalFee = standardFee + Int(ceil(Double(parkTime - standardTime) / Double(unitTime))) * unitFee

    return totalFee
}

func calculateTime(_ x: String, _ y: String) -> Int {

    let xHourToMinute: Int = Int(x[x.startIndex...x.index(x.startIndex, offsetBy: 1)])! * 60
    let xMinute: Int = Int(x[x.index(x.startIndex, offsetBy: 3)...x.index(x.startIndex, offsetBy: 4)])!
    let yHourToMinute: Int = Int(y[y.startIndex...y.index(y.startIndex, offsetBy: 1)])! * 60
    let yMinute: Int = Int(y[y.index(y.startIndex, offsetBy: 3)...y.index(y.startIndex, offsetBy: 4)])!

    return (yHourToMinute + yMinute) - (xHourToMinute + xMinute)
}

//print(calculateTime("18:59", "23:59"))

print(solution([180, 5000, 10, 600], ["05:34 5961 IN", "06:00 0000 IN", "06:34 0000 OUT", "07:59 5961 OUT", "07:59 0148 IN", "18:59 0000 IN", "19:09 0148 OUT", "22:59 5961 IN", "23:00 5961 OUT"]))

//print(solution([120, 0, 60, 591], ["16:00 3961 IN", "16:00 0202 IN", "18:00 3961 OUT", "18:00 0202 OUT", "23:58 3961 IN"]))
